### PR TITLE
ci(atomic): run only affected E2E tests in atomic

### PR DESCRIPTION
**Pull request by y-lakhdar**
This pull request aims to optimize the E2E tests by running only the affected tests.

Identifying only the affected tests by listing the changed files can be tricky in the case of E2E testing. However, since we are testing components in isolation, we can use this approach along with TypeScript AST (Abstract Syntax Tree) to build a tree representation of the atomic code structure. This allows us to create a dependency map between each E2E test file and all the source files that might affect them.

### Rules
* If a change is detected in a source file that is a dependency (direct or indirect) of an E2E test, the dependent test file will be added to the list of tests to run.
* If a change is detected in an E2E test file (or any test-related file), it will be added to the list of tests to run.
* If the changed file has no effect on E2E tests (e.g., updating the README), no tests will be executed.
* If changes are detected in a different package (e.g., /headless), all tests will be executed.
* If an error occurs in the script (e.g., no associated E2E test for a change made in an atomic component), a warning will be thrown and all tests will be executed.

 
### Shard reduction
I also adjusted the number of shards based on the number of results. 1 shard per test suite (24 shard max)

https://coveord.atlassian.net/browse/KIT-3634

### CODE REVIEW

**Review by louis-bompart — COMMENTED** [review]
looking good so far

File: packages/atomic/scripts/list-imports.mjs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        resolveAndAddImport(importPath) ||
+        resolveAndAddImport(join(importPath, 'index')); // Check if the import is from an index file
+
+      if (resolvedFileName) {
+        const fileContent = readFileSync(resolvedFileName, 'utf-8');
+        const sourceFile = getSourceFile(resolvedFileName, fileContent);
+        ts.forEachChild(sourceFile, visit);
+      }
+    }
+  };
```
[the remark this answers is not in the document]
Comment (alexprudhomme) [reply packages/atomic/scripts/list-imports.mjs:64]: It's nice with this we have a circular dependency check

File: scripts/ci/find-tests.mjs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  return searchFiles(dir, []);
+}
+
+/**
+ * Creates a mapping of test file names to the set of files they import.
+ *
+ * @param testPaths - An array of E2E test file paths.
+ * @returns A map where each key is a test file name and the value is a set of files it imports.
```
Comment (fbeaudoincoveo) [inline scripts/ci/find-tests.mjs:46]:
```suggestion
 * @returns A map where each key is a test file name and the value is the set of files it imports.
```

**Review by fbeaudoincoveo — COMMENTED** [review]
Nice!

**Review by louis-bompart — APPROVED** [review]
let's try it!

**Review by alexprudhomme — APPROVED** [review]
nice, ran 1 out of 1 shard 😄

**Pull request merged**